access/ms-teams/app.go

```diff
@@ -0,0 +1,517 @@
+	pd "github.com/gravitational/teleport-plugins/lib/plugindata"
+	"github.com/gravitational/teleport-plugins/lib/stringset"
+	"github.com/gravitational/teleport-plugins/lib/watcherjob"
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+const (
+	// pluginName used as Teleport plugin identifier
+	pluginName = "ms-teams"
+	// minServerVersion is the minimal teleport version the plugin supports.
+	minServerVersion = "6.1.0-beta.1"
+	// grpcBackoffMaxDelay is a maximum time GRPC client waits before reconnection attempt.
+	grpcBackoffMaxDelay = time.Second * 2
+	// minConnectTimeout GRPC connect timeout.
+	minConnectTimeout = time.Second * 10
+	// handlerTimeout is used to bound the execution time of watcher event handler.
+	handlerTimeout = time.Second * 5
+)
+
+// App contains global application state.
+type App struct {
+	conf Config
+
+	apiClient  *client.Client
+	bot        *Bot
+	mainJob    lib.ServiceJob
+	watcherJob lib.ServiceJob
+	pd         *pd.CAS
+
+	*lib.Process
+}
+
+// NewApp initializes a new teleport-ms-teams app and returns it.
+func NewApp(conf Config) (*App, error) {
+	app := &App{conf: conf}
+
+	app.mainJob = lib.NewServiceJob(app.run)
+
+	return app, nil
+}
+
+// Run starts the main job process
+func (a *App) Run(ctx context.Context) error {
+	log := logger.Get(ctx)
+	log.Infof("Starting Teleport MS Teams Plugin %s:%s", Version, Gitref)
+
+	err := a.init(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.Process = lib.NewProcess(ctx)
+	// a.mainJob = lib.NewServiceJob(a.run)
+	a.watcherJob = a.newWatcherJob()
+
+	a.SpawnCriticalJob(a.mainJob)
+	a.SpawnCriticalJob(a.watcherJob)
+	<-a.Process.Done()
+	return a.Err()
+}
+
+// Err returns the error app finished with.
+func (a *App) Err() error {
+	return trace.Wrap(a.mainJob.Err())
+}
+
+// WaitReady waits for http and watcher service to start up
+func (a *App) WaitReady(ctx context.Context) (bool, error) {
+	return a.mainJob.WaitReady(ctx)
+}
+
+// init initializes the application
+func (a *App) init(ctx context.Context) error {
+	bk := grpcbackoff.DefaultConfig
+	bk.MaxDelay = grpcBackoffMaxDelay
+
+	apiClient, err := client.New(ctx, client.Config{
+		Addrs:       a.conf.Teleport.GetAddrs(),
+		Credentials: a.conf.Teleport.Credentials(),
+		DialOpts: []grpc.DialOption{
+			grpc.WithConnectParams(
+				grpc.ConnectParams{
+					Backoff:           bk,
+					MinConnectTimeout: minConnectTimeout,
+				},
+			),
+		},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.apiClient = apiClient
+	a.pd = pd.NewCAS(
+		apiClient,
+		pluginName,
+		types.KindAccessRequest,
+		EncodePluginData,
+		DecodePluginData,
+	)
+
+	pong, err := a.checkTeleportVersion(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var webProxyAddr string
+	if pong.ServerFeatures.AdvancedAccessWorkflows {
+		webProxyAddr = pong.ProxyPublicAddr
+	}
```

Comment:
This logic is not super clear - why are we setting webProxyAddr if AdvancedAccessWorkflows are enabled? What if they're not enabled? The webProxyAddr will be empty and passed to NewBot below - what will happen then?

Reply:
`webProxyAddr` is used to show a fancy "Open" button, which launches Teleport Web UI and opens the access request in question. As far as I understand, that feature exists in Enterprise version only - that's why we check `AdvancedAccessWorkflows`. Yes, empty value indicates to app that it should not render those links. It works like this for all other plugins, should we change it?